Engine/ac/game.cpp

```diff
@@ -269,18 +269,19 @@ void Game_SetAudioTypeVolume(int audioType, int volume, int changeType)
         quitprintf("!Game.SetAudioTypeVolume: volume %d is not between 0..100", volume);
     if ((audioType < 0) || (audioType >= game.audioClipTypeCount))
         quitprintf("!Game.SetAudioTypeVolume: invalid audio type: %d", audioType);
-    int aa;
 
     Debug::Printf("Game.SetAudioTypeVolume: type: %d, volume: %d, change: %d", audioType, volume, changeType);
     if ((changeType == VOL_CHANGEEXISTING) ||
         (changeType == VOL_BOTH))
     {
-        for (aa = 0; aa < MAX_SOUND_CHANNELS; aa++)
+        AudioChannelsLock lock;
+        for (int aa = 0; aa < MAX_SOUND_CHANNELS; aa++)
         {
             ScriptAudioClip *clip = AudioChannel_GetPlayingClip(&scrAudioChannel[aa]);
             if ((clip != NULL) && (clip->type == audioType))
             {
-                channels[aa]->set_volume_percent(volume);
+                auto* ch = lock.GetChannel(aa);
+                ch->set_volume_percent(volume);
```

Comment:
Needs to check if ch is null

Reply:
It's relying on returned information about clip playing on this channel. All this is happening under one lock, so probably safe (at least I would think so).

Reply:
It's relying on scrAudioChannel's information about clip playing on it. There might be more places like this, where engine does not consider that the state may be inconsistent.

Reply:
Oh yeah this is concerning. The existence of scrAudioChannel seems opposed to the safetiness of using my channel locking approach. It's like, more state, for each channel, stored separately. This really needs to be redesigned so it's kept together under control of the audio channel. could all the scrAudioChannel state be stored attached to the channel?

Reply:
@mgambrell this is about something I did in my old attempt few years ago, where I tried to gather all channel-related data under one protected struct. I've mentioned it a bit ago in previous threads:
https://github.com/ivan-mogilko/ags-refactoring/commits/refactor_audio_channel

This was somewhat similar to your approach, except I was not returning clips, but *channels*, which was safer imho because clip may be replaced in the middle of function. For convenience IIRC I overloaded -> operator to access clip through this channel struct.

When these pull requests with threading audio was started to open I suggested to restore that branch, because it already contained lots of work, but then let myself get persuaded to not push too much refactoring now.
I keep getting very dissapointed by myself because of this.

Reply:
The more issues is found in this pr, the more I think that maybe I should restore my old code instead, finish it and add mutexes, using your commit for reference.

Reply:
I mean, this PR is mostly ready. I can help add your channel refactoring after this one if you like.

Reply:
Another point is that unless the AudioChannel is an array of staticly allocated objects, you'll still need to lock the channels array when removing/adding channel pointers.

Reply:
> Another point is that unless the AudioChannel is an array of staticly allocated objects, you'll still need to lock the channels array when removing/adding channel pointers.

AudioChannel actually is the array of statically allocated objects, they are created at engine startup and never changed.

Reply:
> I mean, this PR is mostly ready. I can help add your channel refactoring after this one if you like.

Alright, question is: will additional null-pointer tests you suggested be enough to keep things at least thread safe?

Reply:
I believe so. You might be interested in this commit https://github.com/sonneveld/ags/commit/b2a6fe059cda5dd8be2526766a59f01a96aac8b7 I made some adjustments as I was reviewing this PR.